File: python/problems/leetcode/0051_n_queens/solution.py
"""LeetCode 51: N-Queens."""

from __future__ import annotations


class Solution:
    def solveNQueens(self, n: int) -> list[list[str]]:
        res: list[list[str]] = []
        cols: set[int] = set()
        diag1: set[int] = set()
        diag2: set[int] = set()
        cur: list[int] = []

        def bt(r: int) -> None:
            if r == n:
                board = []
                for c in cur:
                    row = ["."] * n
                    row[c] = "Q"
                    board.append("".join(row))
                res.append(board)
                return
            for c in range(n):
                if c in cols or (r - c) in diag1 or (r + c) in diag2:
                    continue
                cols.add(c)
                diag1.add(r - c)
                diag2.add(r + c)
                cur.append(c)
                bt(r + 1)
                cur.pop()
                cols.discard(c)
                diag1.discard(r - c)
                diag2.discard(r + c)

        bt(0)
        return res


if __name__ == "__main__":
    assert len(Solution().solveNQueens(4)) == 2
    assert Solution().solveNQueens(1) == [["Q"]]
    assert len(Solution().solveNQueens(8)) == 92
    print("0051 OK")

File: cpp/problems/leetcode/0051_n_queens/solution.cpp
// LeetCode 51
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<int> cur;
        vector<bool> cols(n, false), d1(2 * n, false), d2(2 * n, false);
        function<void(int)> bt = [&](int r) {
            if (r == n) {
                vector<string> board(n, string(n, '.'));
                for (int i = 0; i < n; ++i) board[i][cur[i]] = 'Q';
                res.push_back(board);
                return;
            }
            for (int c = 0; c < n; ++c) {
                if (cols[c] || d1[r - c + n] || d2[r + c]) continue;
                cols[c] = d1[r - c + n] = d2[r + c] = true;
                cur.push_back(c);
                bt(r + 1);
                cur.pop_back();
                cols[c] = d1[r - c + n] = d2[r + c] = false;
            }
        };
        bt(0);
        return res;
    }
};

int main() {
    assert((int)Solution().solveNQueens(4).size() == 2);
    assert((Solution().solveNQueens(1) == vector<vector<string>>{{"Q"}}));
    assert((int)Solution().solveNQueens(8).size() == 92);
    cout << "0051 OK" << endl;
    return 0;
}
